packages/medusa/src/services/payment-collection.ts

```diff
@@ -160,4 +253,479 @@ export default class PaymentCollectionService extends TransactionBaseService {
+            select: ["id", "email", "metadata"],
+          })
+          if (!customer) {
+            throw new MedusaError(
+              MedusaError.Types.UNEXPECTED_STATE,
+              `Invalid customer_id`
+            )
+          }
+        }
+
+        const existingSession = payCol.payment_sessions?.find(
+          (sess) => session.session_id === sess?.id
+        )
+
+        const inputData: PaymentProviderDataInput = {
+          resource_id: payCol.id,
+          currency_code: payCol.currency_code,
+          amount: session.amount,
+          provider_id: session.provider_id,
+          customer,
+          metadata: {
+            resource_id: payCol.id,
+          },
+        }
+
+        if (existingSession) {
+          const paymentSession = await this.paymentProviderService_
+            .withTransaction(manager)
+            .updateSessionNew(existingSession, inputData)
+
+          selectedSessionIds.push(existingSession.id)
+          paymentSessions.push(paymentSession)
+        } else {
+          const paymentSession = await this.paymentProviderService_
+            .withTransaction(manager)
+            .createSessionNew(inputData)
+
+          selectedSessionIds.push(paymentSession.id)
+          paymentSessions.push(paymentSession)
+        }
+      }
+
+      if (payCol.payment_sessions?.length) {
+        const removeIds: string[] = payCol.payment_sessions
+          .map((sess) => sess.id)
+          .filter((id) => !selectedSessionIds.includes(id))
+
+        if (removeIds.length) {
+          await paymentCollectionRepository
+            .createQueryBuilder()
+            .delete()
+            .from(PaymentSession)
+            .where("id IN (:...ids)", { ids: removeIds })
+            .execute()
+        }
+      }
+
+      payCol.payment_sessions = paymentSessions
+      await paymentCollectionRepository.save(payCol)
+
+      return payCol
+    })
+  }
+
+  async refreshPaymentSession(
+    paymentCollectionId: string,
+    sessionId: string,
+    sessionInput: PaymentCollectionSessionInput
+  ): Promise<PaymentSession> {
+    return await this.atomicPhase_(async (manager: EntityManager) => {
+      const payCol = await this.getPaymentCollectionIdBySessionId(sessionId, {
+        relations: ["region", "region.payment_providers", "payment_sessions"],
+      })
+
+      if (paymentCollectionId !== payCol.id) {
+        throw new MedusaError(
+          MedusaError.Types.NOT_ALLOWED,
+          `Payment Session ${sessionId} does not belong to Payment Collection ${paymentCollectionId}`
+        )
+      }
+
+      const session = payCol.payment_sessions.find(
+        (sess) => sessionId === sess?.id
+      )
+
+      if (session?.amount !== sessionInput.amount) {
+        throw new MedusaError(
+          MedusaError.Types.NOT_ALLOWED,
+          "The amount has to be the same as the existing payment session"
+        )
+      }
+
+      const customer = await this.customerService_.retrieve(
```

Comment:
todo: Use `withTransaction`
```suggestion
      const customer = await this.customerService_.withTransaction(manager).retrieve(
```